experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,2237 @@
+      : OpExecutionModel<Op>(op), op(op),
+        outputWidth(op.getResult().getType().getDataBitWidth()),
+        callback(callback) {
+    ins = OpExecutionModel<Op>::template getState<ChannelConsumerRW>(
+        op.getOperand(), subset);
+    outs = OpExecutionModel<Op>::template getState<ChannelProducerRW>(
+        op.getResult(), subset);
+  }
+
+  void reset() override {
+    outs->data = callback(ins->data, outputWidth);
+    outs->valid = ins->valid;
+    ins->ready = outs->ready;
+  }
+
+  void exec(bool isClkRisingEdge) override { reset(); }
+
+  void printStates() override {
+    OpExecutionModel<Op>::template printValue<ConsumerRW, const Data>(
+        "ins", ins, &ins->data);
+    OpExecutionModel<Op>::template printValue<ProducerRW, Data>("outs", outs,
+                                                                &outs->data);
+  }
+
+private:
+  Op op;
+
+  // parameters
+  unsigned outputWidth;
+  UnaryCompFunc callback;
+
+  // ports
+  ChannelConsumerRW *ins;
+  ChannelProducerRW *outs;
+};
+
+/// Mutual component for binary operations: AddI, AndI, CmpI, MulI, OrI,
+/// ShlI, ShrsI, ShruI, SubI, XorI...
+using BinaryCompFunc = std::function<Data(const Data &, const Data &)>;
+template <typename Op>
+class GenericBinaryOpModel : public OpExecutionModel<Op> {
+public:
+  using OpExecutionModel<Op>::OpExecutionModel;
+  GenericBinaryOpModel(Op op, mlir::DenseMap<Value, RW *> &subset,
+                       const BinaryCompFunc &callback, unsigned latency = 0)
+      : OpExecutionModel<Op>(op), op(op), callback(callback), latency(latency),
+        bitwidth(cast<handshake::ChannelType>(op.getResult().getType())
+                     .getDataBitWidth()),
+
+        lhs(OpExecutionModel<Op>::template getState<ChannelConsumerRW>(
+            op.getLhs(), subset)),
+        rhs(OpExecutionModel<Op>::template getState<ChannelConsumerRW>(
+            op.getRhs(), subset)),
+        result(OpExecutionModel<Op>::template getState<ChannelProducerRW>(
+            op.getResult(), subset)),
+        outsJoin(joinValid, result->ready), binJoin(2) {
+    insJoin = {lhs, rhs};
+  }
+
+  void reset() override {
+    binJoin.exec(insJoin, result);
+    result->data = callback(lhs->data, rhs->data);
+  }
+
+  void exec(bool isClkRisingEdge) override {
+    if (!latency)
+      reset();
+    else {
+      binJoin.exec(insJoin, &outsJoin);
+
+      if (isClkRisingEdge) {
+        if (lhs->valid && rhs->valid && !hasData) {
+          // collect
+          hasData = true;
+          counter = 1;
+          tempData = callback(lhs->data, rhs->data);
+          result->valid = false;
+        } else if (hasData && counter < latency - 1) {
+          // wait
+          ++counter;
+          if (counter == latency - 1)
+            result->data = tempData;
+        } else if (hasData && counter >= latency - 1 && !result->valid) {
+          result->data = tempData;
+          result->valid = true;
```

Comment:
It seems that this `if` and `else if` sort of mean the same thing. In both cases you set the data but only in one of the two you set the valid, which is slightly weird to me (the data only needs to be set when you set the result to valid, before it is assumed to be "garbage"). Can you make sure this is correct and attempt to rewrite it so there there is a single place where you set the data and valid to `true`?